broadinstitute/single_cell_portal_core
# Handle many long, odd labels in custom scatter plot legend (SCP-3818)

This refines the new custom group scatter plot legend to handle legends that have labels that are long or many in number.  It takes this by using hiding text that would overflow behind an ellipsis (...), and showing the full text upon hovering.  It also visually compresses related content like "(52 points)" to simply "52", and likewise showing more context on hover.

Here's a video of how it looks:

https://user-images.githubusercontent.com/1334561/140769939-c40a795e-4efb-4634-b1a8-39adb00c79ab.mov

To accommodate label sets of cells that have potentially many cells, while ensuring the number of points is always displayed, labels for sets with _few_ cells are often truncated before they would overlap.  The fixed width handles sets with up to 7 digits (i.e. single-digit-millions of cells).

Automatically testing such layout handling is complicated, so I've added a synthetic dataset to ease manually testing.

To test:
* Populate "mouse_salmonella" study, e.g. via `echo "SyntheticStudyPopulator.populate('mouse_salmonella')" | bundle exec rails console -e development`.
* Navigate to it in web browser
* In "Clustering" dropdown menu, select "cluster_many_long_odd_labels.tsv"
* Verify long labels do not overflow the legend, and each are contained in one line
* Hover over a truncated label
* Verify full label appears nearby in small box
* Hover over number at right in a label
* Verify it shows the number of points with more context

This satisfies SCP-3818.
